infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/EnvironmentVariableSecretApplier.java

```diff
@@ -0,0 +1,110 @@
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment.PodData;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment.PodRole;
+
+/**
+ * Mounts Kubernetes secret with specific annotations as an environment variable(s) in workspace
+ * containers. Allows per-component control of secret applying in devfile.
+ */
+@Beta
+@Singleton
+public class EnvironmentVariableSecretApplier
+    extends KubernetesSecretApplier<KubernetesEnvironment> {
+
+  static final String ANNOTATION_ENV_NAME = ANNOTATION_PREFIX + "/" + "env-name";
+  static final String ANNOTATION_ENV_NAME_TEMPLATE = ANNOTATION_PREFIX + "/%s_" + "env-name";
+
+  @Override
+  public void applySecret(KubernetesEnvironment env, Secret secret) throws InfrastructureException {
+    boolean secretAutomount =
+        Boolean.parseBoolean(secret.getMetadata().getAnnotations().get(ANNOTATION_AUTOMOUNT));
+    for (PodData podData : env.getPodsData().values()) {
+      if (!podData.getRole().equals(PodRole.DEPLOYMENT)) {
+        continue;
+      }
+      for (Container container : podData.getSpec().getContainers()) {
+        Optional<ComponentImpl> component = getComponent(env, container.getName());
+        if ((component.isPresent() && isOverridenByFalse(component.get()))
+            || (!secretAutomount
+                && !(component.isPresent() && isOverridenByTrue(component.get())))) {
+          continue;
+        }
+        for (Entry<String, String> secretDataEntry : secret.getData().entrySet()) {
+          final String mountEnvName = envName(secret, secretDataEntry.getKey());
+          container
+              .getEnv()
+              .add(
+                  new EnvVarBuilder()
+                      .withName(mountEnvName)
+                      .withValueFrom(
+                          new EnvVarSourceBuilder()
+                              .withSecretKeyRef(
+                                  new SecretKeySelectorBuilder()
+                                      .withName(secret.getMetadata().getName())
+                                      .withKey(secretDataEntry.getKey())
+                                      .build())
+                              .build())
+                      .build());
+        }
+      }
+    }
+  }
+
+  private String envName(Secret secret, String key) throws InfrastructureException {
+    String mountEnvName;
+    if (secret.getData().size() == 1) {
+      try {
+        mountEnvName =
+            firstNonNull(
+                secret.getMetadata().getAnnotations().get(ANNOTATION_ENV_NAME),
+                secret
+                    .getMetadata()
+                    .getAnnotations()
+                    .get(format(ANNOTATION_ENV_NAME_TEMPLATE, key)));
+      } catch (NullPointerException e) {
```

Comment:
from `firstNonNull()`. Another way is to assign to variables and check if they are both `null` then thwor exception. So utility method does that more effective and less computations.

Reply:
I'm not a big fan of NPE exception handling as the main logical tree. @sparkoo @metlos opinions?

Reply:
it's not main logical branch. Its exceptional situation when further execution shuld be aborted. So i see no problem.

Reply:
A safe code should not throw NPE.
=)

firstNonNull is supposed to be called with AT LEAST non-null argument.
I think It's better to avoid exception throwing/catching in non-exceptional situations. 

So, here is my +1 to avoid it here

Reply:
Why do you decide this as non-exceptional situation ? i cannot understood..

Reply:
> A safe code should not throw NPE.
This code _never_ thows an NPE.
> firstNonNull is supposed to be called with AT LEAST non-null argument.
Not really true (or where do you read that?). Its just a question if the NPE thrown by this method is suitable. If it is supposed to continue execution of the branch, then i agree, NPE is not suitable. 

 So to stop flame i'll rewrite it on streams, but still think this is quite acceptable code :- )